src/paging.rs

```diff
@@ -539,6 +555,35 @@ impl<T: Translation> PageTableWithLevel<T> {
         }
     }
 
+    /// Modifies a range of page table entries by applying a function to each page table entry.
+    fn modify_range(
+        &mut self,
+        translation: &T,
+        range: &MemoryRegion,
+        f: &PteUpdater,
+    ) -> Result<(), MapError> {
+        let level = self.level;
+        for chunk in range.split(level) {
+            let base_va = chunk.0.start.0;
```

Comment:
The updater function is called once for every descriptor which maps any part of the range passed to `modify_range`, regardless of alignment. So, if that range starts halfway through a block mapping, the function will still be called once.

The range passed as the first argument of the updater function is the entire range mapped by the page table entry (the second argument), without any clamping to start/end of the original range passed to `modify_range`.

Reply:
That seems confusing, as you then end up updating a larger range than was requested. Would you ever want to do that? How about checking and returning an error instead if the range is not properly aligned with existing block mappings? I think you only need to check the beginning and end, so it could be done in `RootTable::modify_range`.

Reply:
Actually, looking at how `MemoryRegion::split` works, I don't think it's even true that the whole range covered by the entry is passed to the updater function. The first region returned by `split` will start at the same address as the original region, which might be partway through a block mapping. When you construct the new `MemoryRegion` below to pass to the update function, you add the granularity to it, so it may end up overlapping the next block, which is definitely wrong.